### DISCUSSION TITLE
Istio Ingress Gateway health checks against target destination

### DISCUSSION THREAD

**Discussion by davidgwps** [istio/istio#52086] [Q&A]
Hey, I'm looking for some advice on Istio Ingress Gateway and if it is possible to customise our L7 load balancers health check to not just check the health of the gateway, but also the target destination.

In the case where you have a backend, with endpoints in 2 different regions, at the moment with a health check on `:15021/healthz/ready` if you take down the replicas of your service (not Istio Gateway) in one region, you'll get a mix of 503's and 200's depending on what it hits. I believe this is because the Istio Gateway is healthy, but the target pods are unavailable.

Is there a way to accomplish this, and remove the backend that is unhealthy, without meshing the two regions together?

**Comment by BeyondEvil** [comment]
Did you ever figure this out?

**Comment by rsalmond** [comment]
Different cloud providers do load balancing slightly differently, but in general the health check between an LB and an ingress gateway deployment is there so that the LB can decide which ingress gateway pods are ready to receive traffic. An ingress gateway pod can potentially route traffic to any number of different deployments in the cluster. Say there are 2 deployments Foo and Bar in the cluster and the ingress gateway routes /foo to one and /bar to the other. 

Now say that Bar has a problem and begins to fail. If the LB is health checking Foo and Bar (instead of the ingress gateway itself) it will now mark all pods in the ingress gateway as unhealthy because Bar is unhealthy. This will result in all traffic to Foo, which is fine, failing.

**Reply by BeyondEvil** [reply]
Makes sense. So in short: You want the LB to health check against the gateway and not the services behind it?

**Reply by rsalmond** [reply]
Exactly. Once the request is inside the cluster, Istio has its own features which can stop traffic to a single misbehaving pod (outlier detection) or shift traffic over to another cluster if the local deployment falls over (locality failover).

**Reply by HristoHT** [reply]
So Imagine you have 2 K8s clusters in different regions - region A and region B. How to detect if a region A is down and redirect the Top level LB to region B if ingress gateway in region A is always healthy :D

**Reply by rsalmond** [reply]
AFAIK there is no built-in mechanism for multicluster health checking in kubernetes, although I am not up to date on SIG-multicluster work so if there's something relevent there please let me know.

**Reply by HristoHT** [reply]
I don't understand you. 
Here some context. My application uses CloudFlare DNS resolving to different ELBs in different regions. Each ELB target health group is not the downstream application pods but the istio ingress gateways. So how CloudFlare checks the health status of the cluster? - CloudFlare pings ELB - ELB pings Istio Ingress Gateways - they say healthy. (oversimplified for easier understanding)
On CloudFlare level I can failover to different regions when we have indication that the ELB for a region is unhealthy (ideally when the application pods are unhealthy, not the istio ingress gateway pods) But currently we cannot propagate some status of the healthiness of the underlying pods.

Here I try to explain visually how the "health check flow" differs from the "networking flow". 

PS. ELB is Elastic Load Balancer (some AWS stuff)
PSPS. This is edge case where outage happens in our cluster causing all application service pods to die but the istio ingress gateways are okay" 

<img width="1200" height="816" alt="Screenshot 2025-11-10 at 17 20 08" src="https://github.com/user-attachments/assets/8068075b-05b4-47be-8606-84b9d3cd6613" />

**Reply by BeyondEvil** [reply]
If the Gateway is healthy, but has no healthy services, it will respond with "no healthy upstream" (Unsure what the HTTP code is).

**Reply by HristoHT** [reply]
Thank you for the insight this will be also useful @BeyondEvil 🙌
 
But ideally we want to detect outage and trigger automated failover to other region before all pods are dead.  And I think this is common sense to be done via the health checks. Because the cluster is not healthy as long as the istio ingress gateways are healthy :D

**Reply by BeyondEvil** [reply]
> But ideally we want to detect outage and trigger automated failover to other region before all pods are dead.

Why?

**Reply by HristoHT** [reply]
> Why?

Short answer - availability of our service.

Imagine we are using local for the region MySQL and something happened to it (like transactions locks a table) and and apps health checks detect that and we want to failover to other region while somebody investigates and fix the issue in the current region.

**Reply by BeyondEvil** [reply]
Right, but I still don't understand why you want to switch over when there's still _healthy_ pods?

**Reply by HristoHT** [reply]
It is business requirement to have highly available system. And automatically redirecting part of the traffic to other more stable region.

Or I don't understand your question, could you please elaborate more?

**Reply by BeyondEvil** [reply]
“HA” and “business requirements” don’t mean much without hard targets, you need actual numbers (SLOs/SLAs) to design against.

Assuming the goal is to let the ALB determine whether a region or cluster is healthy (and assuming you use a shared Istio Gateway, not one per workload), here’s how I’d approach it:

- Deploy a small probe service that continuously runs synthetic checks through the same Gateway paths real clients use.
- It tracks success rate and p95 latency (and any other SLO metrics that matter).
- If those fall outside thresholds for a short rolling window, the probe reports unhealthy.
- Point the ALB health check at this probe endpoint instead of the Gateway’s own readiness probe.
- That way, your upstream load balancer measures actual user-perceived health, not just “is the Gateway Pod up.”

Hope that makes sense.
